### ISSUE TITLE
Feature Extraction

### ISSUE THREAD

**Issue by aleflabo** [assembly-101/assembly101-action-recognition#11]
Hi! 

I'm trying to extract the features of a dataset after training the proposed TSM on it. Following the suggestion in the issue https://github.com/assembly-101/assembly101-action-recognition/issues/1, I should train the 2D backbone without shift, and this is what I'm trying:

`python main.py dataset RGB --arch resnet50 --num_segments 8 --gd 20 --lr 0.001 --lr_steps 20 40 --epochs 50 --batch-size 32 -j 16 --dropout 0.5 --consensus_type=avg --eval-freq=1 --npb --tune_from=pretrained_models/tsm_rgb_epic.ckpt --gpus 0 1
`
In this way the features in output of the base_model have a shape of (128, 2048). This is due to the number of segments, 128 = num_segments*batch_size. 

How should I handle this problem to save the correct 2048-D tensor per frame?

Kind regards,
Alessandro

_Originally posted by @aleflabo in https://github.com/assembly-101/assembly101-action-recognition/issues/1#issuecomment-1791697133_

**Comment by dibschat** [comment]
Hi Alessandro,

The TSM is trained using the usual 8/16-frame input strategy (with temporal shifts) on Assembly101. We only extract the per-frame TSM features using the 2D backbone of the pretrained TSM (from the first sentence), without any shifting operation.

Please feel free to reply to this if you have any further queries.

Best,
Dibyadip

**Comment by aleflabo** [comment]
Got it, thank you! 

Kind regards,
Alessandro

**Comment by aleflabo** [comment]
Hi Dibyadip,

I've followed your suggestions and extracted the features after the 2D backbone. Precisely, the features are saved after the forward of the base_model (at this line: https://github.com/assembly-101/assembly101-action-recognition/blob/00d3c3bb3af75ce38ae406614a7cb034aecdebc0/TSM-action-recognition/ops/models.py#L272). 

All the features saved have the same indexes with the value 0. Here is an example of the first 20 features (out of 2048) of a frame and the zeros in different positions can be seen. 
```
array([0.        , 0.00726067, 0.06628822, 0.        , 0.01285915,
       0.01205407, 0.00071825, 0.23438536, 0.        , 0.        ,
       0.        , 0.00035547, 0.00821526, 0.        , 0.        ,
       0.        , 0.        , 0.02248539, 0.06791941, 0.        ],
      dtype=float32)
```

I think this is happening cause the shifting is not enabled. Do you have any advice on how to handle this problem?
The train is done using the command I posted in the first comment. 

Thanks,
Alessandro

**Comment by aleflabo** [comment]
The last layer of the base_model is a Dropout layer and that seems to be causing the zeros in the output. 
Should I just replace it with an Identity layer?

The last two layers of the base_model are: 
```
(avgpool): AdaptiveAvgPool2d(output_size=1)
(fc): Dropout(p=0.8, inplace=False)
```

Looking at the code of RULSTM 
(https://github.com/fpv-iplab/rulstm/blob/e648c7ee40a1a71381bfaa7563f2e8a6935275ab/FEATEXT/extract_example_rgb.py#L22C9-L22C9
), they replace the last two layers. Should I do the same?


Thanks,
Alessandro

**Comment by dibschat** [comment]
Hi Alessandro, 

Yes, as mentioned in issue #6, we follow RU-LSTM extraction code by just replacing TSN with TSM and also changing the input transforms. This means replacing the dropout layer with `nn.Identity()` i.e. `model.base_model.fc = nn.Identity()`

Please feel free to open this issue if you have any further queries.

Best,
Dibyadip
